### ISSUE TITLE
Incorporation of PSF in Model 

### ISSUE THREAD

**Issue by JonnyyTorres** [JonnyyTorres/Galsim_JAX#5]
Hello @b-remy and @EiffL. As we discused, the [last code version](https://github.com/JonnyyTorres/Galsim_JAX/blob/b1c4cc3b61657aa97647dd5e863c920cb8608d4d/VAE_SD_C.py) contains the convolution with the PSF in the loss function by using the COSMOS Dataset. Here, I'd like to present you some interesting results to know your opinion and the steps we should test. First of all, some outputs of the (Raw) Autoencoder with the SD architecture that achieved the best log-likelihood results to date: 

![Autoencoder_samples](https://github.com/JonnyyTorres/Galsim_JAX/assets/92088528/f50ee6a7-9afb-42b1-807a-241ef2337351)

The same images but convolved with the PSF:

![Convolve_samples](https://github.com/JonnyyTorres/Galsim_JAX/assets/92088528/cfd9cbcb-88e7-486d-ad1a-8c37cd8a3cc8)

And finally the same examples but this time taking the convolved images as the mean of a Multivariate Normal Diagonal Distribution and then sampled from this distribution 

![MNDist_samples](https://github.com/JonnyyTorres/Galsim_JAX/assets/92088528/50aa0a74-831a-4a9a-8c31-b6c2492eec40)

As you can appreciate the added noise in the output is quite appreciable. In general, the magnitude of the galaxies in this dataset as you know is higher, being reflected in a lower SNR compared to the HSC dataset. 

This week, I have been waiting for the variance calculation for each image in the dataset. It took a considerable time, but I hope to be able to incorporate it in the loss function soon in order to observe the changes.

I remain attentive to your observations and comments :-)

**Comment by JonnyyTorres** [comment]
Hello @b-remy and @EiffL. Today I would like to present you some progress on the model with a better loss after trying different configurations. First the images obtained by the autoencoder: 

![Autoencoder_samples(1)](https://github.com/JonnyyTorres/Galsim_JAX/assets/92088528/26673f2c-d517-4919-8684-af4da6e469cc)

As you can see, I still have some diagonal stripes (with Benjamin we lowered the hypothesis that it may be due to the way the convolution was performed). Here are the images after convolution with the PSF:

![Convolve_samples(1)](https://github.com/JonnyyTorres/Galsim_JAX/assets/92088528/44404304-83b3-40f4-bf3e-a97575b31048)

And finally the images after being sampled from the Diagonal Multivariate Gaussian Diagonal Distribution:

![MNDist_samples(1)](https://github.com/JonnyyTorres/Galsim_JAX/assets/92088528/b4bd4fb7-d433-40d5-aa3a-505e1ffba6fb)

As you can see, the results are starting to look a little better, although I suppose they are still susceptible to improvement, either by incorporating a total variance factor or by modifying the convolution function. 

On the other hand, I share with you this gif where you can appreciate the difference between the original images and the resulting ones after going through the autoencoder :-) :

![ezgif com-crop](https://github.com/JonnyyTorres/Galsim_JAX/assets/92088528/dc7d2936-e251-4f8b-bcb9-6755919d4865)

The hyperparameters I used to obtain this result are listed below:

Key | Value
-- | --
act_fn | "gelu"
batch_size | 16
interpolation | "Bicubic"
kl_reg | 0.001
latent_dim | 128
learning_rate | 0.01
opt | "adafactor"
resnet_blocks | 2
steps | 125,000
steps_schedule | 50,000

As for the Autoencoder architecture, it is still the same as the one I usually use, that is, three symmetrical convolution layers for the Encoder and the Decoder, having two ResNet Blocks for each Layer. At the moment, the latent space dimension I am working with is **32²**, however, Benjamin tells me that it might be more convenient to work with a 16² dimension, test that I will perform today to see if I get some improvements in the loss values.

**Comment by b-remy** [comment]
I think we're good with the convolution now

![difference_pred_125005_f8251ddc4554c2d1215c.png](https://github.com/JonnyyTorres/Galsim_JAX/assets/30293694/e86e192f-68ca-43b8-9191-1b350ff5b8d0)

Let's see with your best training setting @JonnyyTorres :-)
